### PR TITLE
fix(core): fix re-insertions in the iterable differ

### LINKED ISSUE #17852: NgClass removes existing class when applying array that contains with 2 empty slots
## I'm submitting a ...

<pre><code>
[ ] Regression (behavior that used to work and stopped working in a new release)
[x] Bug report 
[ ] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
</code></pre>

**Construct a class array with 4 strings. The middle classes are an empty string. Then change the 2nd class, leaving the other classes unchanged.**

## Current behavior

NgClass does not apply correct classes. It removes the last class from the array. Internally the last class is added to the add and remove list, and since the add class code does nothing since it's already in the list and does not error, the remove list removes the class.

## Expected behavior

NgClass apply correct classes. The action should be to do nothing for the other existing classes and only add the 2nd class.

The following code tries to add the existing class in `changes.forEachAddedItem`, then removes it in `changes.forEachRemovedItem`. I would suggest not only fixing the bug, but reverse the order to first remove, then add, in case there are more bugs in the future.

https://github.com/angular/angular/blob/master/packages/common/src/directives/ng_class.ts#L105
```
private _applyIterableChanges(changes: IterableChanges<string>): void {
    changes.forEachAddedItem((record) => {
      if (typeof record.item === 'string') {
        this._toggleClass(record.item, true);
      } else {
        throw new Error(
            `NgClass can only toggle CSS classes expressed as strings, got ${stringify(record.item)}`);
      }
    });

    changes.forEachRemovedItem((record) => this._toggleClass(record.item, false));
}
```

## Minimal reproduction of the problem with instructions

http://plnkr.co/edit/bTsnyx1EciH66tRROu5o?p=preview

## What is the motivation / use case for changing the behavior?

Developers expect binding to be correct.

## Please tell us about your environment

<pre><code>
Angular version: 4.2.4

Browser:
- [x] Chrome (desktop) version 59
- [ ] Chrome (Android) version XX
- [ ] Chrome (iOS) version XX
- [ ] Firefox version XX
- [ ] Safari (desktop) version XX
- [ ] Safari (iOS) version XX
- [ ] IE version XX
- [x] Edge version 15
 
For Tooling issues:
- Node version: XX 
- Platform: 

Others:

</code></pre>

**Updated all NgStyle to NgClass** - sorry for the brain fart.

### PR DESCRIPTION
**Pull request by vicb**
fixes #17852

### CODE REVIEW

File: packages/core/src/change_detection/differs/default_iterable_differ.ts [hunk 1]
```diff
@@ -669,7 +670,7 @@ class _DuplicateItemRecordList<V> {
   get(trackById: any, afterIndex: number|null): IterableChangeRecord_<V>|null {
     let record: IterableChangeRecord_<V>|null;
     for (record = this._head; record !== null; record = record._nextDup) {
-      if ((afterIndex === null || afterIndex < record.currentIndex !) &&
+      if ((afterIndex === null || afterIndex <= record.currentIndex !) &&
```
Comment (vicb) [inline packages/core/src/change_detection/differs/default_iterable_differ.ts]: `=` is the fix

File: packages/core/test/change_detection/differs/default_iterable_differ_spec.ts [hunk 2]
```diff
@@ -294,6 +281,22 @@ export function main() {
         }));
       });
 
+      // https://github.com/angular/angular/issues/17852
+      it('support re-insertion', () => {
```
Comment (vicb) [inline packages/core/test/change_detection/differs/default_iterable_differ_spec.ts]: and here is the test

File: packages/core/src/change_detection/differs/default_iterable_differ.ts [hunk 3]
```diff
@@ -669,7 +670,7 @@ class _DuplicateItemRecordList<V> {
   get(trackById: any, afterIndex: number|null): IterableChangeRecord_<V>|null {
```
Comment (chuckjaz) [inline packages/core/src/change_detection/differs/default_iterable_differ.ts]: consider changing the name of `afterIndex` to `atOrAfterIndex` as the meaning of this parameter has changed.

**Review by chuckjaz — APPROVED** [review]
NIT.

**Review by IgorMinar — APPROVED** [review]

**Pull request merged**